tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs

```diff
@@ -170,15 +170,18 @@ public void InjectPathwayContextAsBase64String<TCarrier>(PathwayContext? context
         try
         {
             var previousContext = parentPathway;
-            if (previousContext == null && PreviousCheckpoint.Value != null && PreviousCheckpoint.Value.Kind != checkpointKind)
+            if (previousContext == null && LastConsumePathway.Value != null && checkpointKind == CheckpointKind.Produce)
             {
-                previousContext = PreviousCheckpoint.Value.Context;
+                // We only enter here on produce: when we consume, the only thing that matters is the parent we'd have read from the inbound message, not what happened before.
+                // We want to use the context from the previous consume (but we'll give priority to the parent passed in param if set).
+                previousContext = LastConsumePathway.Value;
             }
 
             var nowNs = DateTimeOffset.UtcNow.ToUnixTimeNanoseconds();
             // We should use timeInQueue to offset the edge / pathway start if this is a beginning of a pathway
             // This allows tracking edge / pathway latency for pipelines starting with a queue (no producer instrumented upstream)
             // by relying on the message timestamp.
+            // ReSharper disable once ArrangeRedundantParentheses
             var edgeStartNs = previousContext == null && timeInQueueMs > 0 ? nowNs - (timeInQueueMs * 1_000_000) : nowNs;
```

Comment:
added this because it kept on removing the parenthesis around `timeInQueueMs * 1_000_000`